ttyplot.c

```diff
@@ -289,9 +290,99 @@ void redraw_screen(const char * errstr) {
     refresh();
 }
 
+// Handle a single value from the input stream.
+// Return whether we got a full data record.
+int handle_value(double value)
+{
+    static double saved_value;
+    static int saved_value_valid = 0;
+
+    // First value of a 2-value record: save it for later.
+    if (two && !saved_value_valid) {
+        saved_value = value;
+        saved_value_valid = 1;
+        return 0;
+    }
+
+    // Otherwise we have a full record.
+    n = (n+1) % plotwidth;
+    if (two) {
+        values1[n] = saved_value;
+        values2[n] = value;
+        saved_value_valid = 0;
+    } else {
+        values1[n] = value;
+    }
+    if (rate)
+        td = derivative(&values1[n], two ? &values2[n] : NULL, &now);
+    return 1;
+}
+
+// Parse the input data. Return the number of bytes consumed.
+size_t parse_input(char *buffer, size_t length)
+{
+    static const char delimiters[] = " \t\r\n";  // white space
+
+    // Find the last delimiter.
+    int end;
+    for (end = length - 1; end >= 0; end--) {
+        if (strchr(delimiters, buffer[end]))  // found
+            break;
+    }
+    if (end < 0)  // no delimiter
+        return 0;
+    buffer[end] = '\0';
+
+    // Tokenize and parse.
+    int records = 0;  // number or records found
+    char *str = buffer;
+    char *token;
+    while ((token = strtok(str, delimiters)) != NULL) {
+        records += handle_value(strtod(token, NULL));
+        str = NULL;  // tell strtok() to stay on the same string
+    }
+    v += records;
+    if (records > 0)
+        redraw_screen(errstr);
```

Comment:
If `parse_input` is about parsing input, I would like to argue that it calling `redraw_screen` is out of place here because it violates the contract of what is the job of function `parse_input`. Can we pull that call out some way? I believe all calls to `redraw_screen` would ideally be from the `main` function.

Reply:
Maybe `parse_input` has just a bad name. I intended for the function to handle the chunk of input data we just grabbed. This means parsing, storing the numbers in the ring buffers and, if needed, calling `redraw_screen`.

I just renamed the function to better reflect their responsibilities:
* `handle_input` → `handle_input_event`
* `parse_input` → `handle_input_data`

> I believe all calls to `redraw_screen` would ideally be from the `main` function.

I don't think they belong there. The purpose of this refactoring is to adhere to the conventional event-driven paradigm. There, the main loop does nothing other than catching and dispatching events. The event handler for the "input ready" event is responsible for _fully_ handling the event. “Fully” means that, once it's handled, we can go to sleep and wait for the next event.

That being said, I left the handling of the input being closed in the main loop, as `stdin_is_open` is local to main. I could move `redraw_screen` there too, but then `handle_input()` would need a way to communicate that a redraw is required. Maybe a global `redraw_needed` boolean? Not sure I like that better. For me, it makes sense to say that redrawing is part of the job of handling the input... if that input implies that a redraw is needed.

Reply:
@edgar-bonet I hope I can concince you still that fewer duties — a smaller "scope" — for these two functions is beneficial and that calling `redraw_screen` is just outside of their ideal scope: they already do "enough".  It should not be hard to pull the drawing call out by the way, e.g.:

- For function `handle_input_data`, we could compare the value of `v` before and after the call and re-draw whenever `v_backup` differs from `v` or something like that.  Or we add a new output parameter `bool * records_written` and inspects its value, to add an alternative.  Note how name `records_written` does not mention drawing, which would be consistent with the view that drawing is not that functions' business, even in their contract with their outside world.

- For function `handle_input_event` after the call we could check if `errstr` has now become non-`NULL` to know if a redraw is needed.

I should add that currently `handle_input_event` seems  missing one more call to `redraw_screen` right after line `errstr = strerror(errno);`.  Pulling redraw out of those two functions, we may also be able to turn three calling places into a single one.  I would be very happy then :smiley:

What do you think?

Reply:
Not that I'm fan of it but... I moved the redraw to a single place of the main loop.

Reply:
@edgar-bonet the increase in global state is not ideal and this way drawing is still among those functions businesses… but turning more global state into passing things around is maybe better a topic for another time. Thanks for this adjustment.